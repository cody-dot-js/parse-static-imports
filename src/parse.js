import parseModuleName from "./parseModuleName";
import importLike from "./importLike";
import parseStarImport from "./parseStarImport";
import parseDefaultImport from "./parseDefaultImport";
import parseNamedImports from "./parseNamedImports";
import isSideEffectOnly from "./isSideEffectOnly";

export default function parse(file = "") {
  return file
    .split("import")
    .map(str => `import ${str}`)
    .filter(importLike)
    .map(str => str.trim())
    .reduce((acc, str) => {
      const sideEffectOnly = isSideEffectOnly(str);
      const imports = sideEffectOnly
        ? {}
        : {
            starImport: parseStarImport(str),
            namedImports: parseNamedImports(str),
            defaultImport: parseDefaultImport(str)
          };
      return [
        ...acc,
        {
          moduleName: parseModuleName(str),
          ...imports,
          sideEffectOnly
        }
      ];
    }, []);
}
