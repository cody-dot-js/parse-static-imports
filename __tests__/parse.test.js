import fs from "fs";
import path from "path";
import { promisify } from "util";
import parse from "../src/parse";
import importSyntaxList from "./__fixtures__/importSyntaxList";

const readFileAsync = promisify(fs.readFile);
const fixturesDirectoryPath = path.resolve(__dirname, "__fixtures__");

test("when given nothing returns an empty list", () => {
  const expected = [];
  const result = parse();

  expect(result).toStrictEqual(expected);
});

describe("for each example in the import syntax list it returns the expected shape", () => {
  importSyntaxList.forEach(({ actual, expected }) => {
    const result = parse(actual);

    it(actual, () => {
      expect(result).toStrictEqual(expected);
    });
  });
});

test("it matches the snapshot for each app example fixture", async () => {
  const appExamplesPath = path.resolve(fixturesDirectoryPath, "app-examples");

  const appExamplePaths = fs
    .readdirSync(appExamplesPath)
    .map(filename => path.join(appExamplesPath, filename));

  const appExamples = await Promise.all(
    appExamplePaths.map(fixturePath => readFileAsync(fixturePath, "utf8"))
  );

  appExamples.forEach((fixture, index) => {
    const result = {
      result: parse(fixture),
      name: path.parse(appExamplePaths[index]).base
    };

    expect(result).toMatchSnapshot();
  });
});

test("when given gibberish it returns an empty list", () => {
  const expected = [];
  const gibberish =
    "Esse adipisicing ad magna sunt exercitation deserunt labore.";
  const result = parse(gibberish);

  expect(result).toStrictEqual(expected);
});
