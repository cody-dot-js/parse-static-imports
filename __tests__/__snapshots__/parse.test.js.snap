// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot for each fixture 1`] = `
Object {
  "fixtureName": "App.jsx",
  "result": Array [
    Object {
      "defaultImport": "React",
      "moduleName": "react",
      "namedImports": Array [
        Object {
          "alias": "foo",
          "name": "useState",
        },
        Object {
          "alias": "useCallback",
          "name": "useCallback",
        },
        Object {
          "alias": "useContext",
          "name": "useContext",
        },
        Object {
          "alias": "useEffect",
          "name": "useEffect",
        },
        Object {
          "alias": "useLayoutEffect",
          "name": "useLayoutEffect",
        },
        Object {
          "alias": "useMemo",
          "name": "useMemo",
        },
        Object {
          "alias": "useRef",
          "name": "useRef",
        },
        Object {
          "alias": "useReducer",
          "name": "useReducer",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "react-dom",
      "namedImports": Array [
        Object {
          "alias": "render",
          "name": "render",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./logo.svg",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "logos",
    },
    Object {
      "moduleName": "./App.css",
      "sideEffectOnly": true,
    },
  ],
}
`;

exports[`it matches the snapshot for each fixture 2`] = `
Object {
  "fixtureName": "index.jsx",
  "result": Array [
    Object {
      "defaultImport": "React",
      "moduleName": "react",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "react-dom",
      "namedImports": Array [
        Object {
          "alias": "render",
          "name": "render",
        },
      ],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "moduleName": "./index.css",
      "sideEffectOnly": true,
    },
    Object {
      "defaultImport": "App",
      "moduleName": "./App",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "",
    },
    Object {
      "defaultImport": "",
      "moduleName": "./serviceWorker",
      "namedImports": Array [],
      "sideEffectOnly": false,
      "starImport": "serviceWorker",
    },
  ],
}
`;
